### ISSUE TITLE
Better display ethereum function calls

### ISSUE THREAD

**Issue by ligi** [trezor/trezor-core#188]
Currently when calling ethereum functions you only see the raw hex data - also often you do not see all of it as the hex gets long fast and I was not able to scroll.
The hex can be easily made shorter and more digestible by humans by resolving the function from the 4byte signature - e.g. via https://github.com/ethereum-lists/4bytes and then displaying the function name with parameters. So a really long hex string that might not fit on the trezor and is hard to parse by a human can easily fit on the display of the trezor and better digested by humans.
A problem might be the size of the signature database - I see 2 solutions here:
 - leverage the SD-Card
 - reduce the database to commonly used signatures only - we can parse all transactions currently done on chain - sort the 4bytes by usage and define a cut-off point

**Comment by prusnak** [comment]
I am planning to create a mechanism which will allow a wallet to upload a signed definition of {coin, token, 4byte-function} as a part of signed transaction.

We would need to create a signed database where each row is signed by trusted key (public key of this keypair will be part of the Trezor firmware) and TREZOR will parse and check the signature of provided data-row.

That way, we would not need to store all coins/tokens/4bytes in the firmware.

**Comment by ligi** [comment]
That is a great idea! Please keep me updated on this

**Comment by ligi** [comment]
I do not see many collisions - I was expecting more and was surprised that there where not too many collisions. In the 23504 signatures I parse currently there are 3 collisions:

aaaaaad1 yycU4()
aaaaaad1 uYZeB()

00000000 get_block_hash_257335279069929(uint256)
00000000 overdiffusingness(bytes,uint256,uint256,uint256,uint256)
00000000 left_branch_block(uint32)

0000006e bright_peace(bytes32,bytes)
0000006e display_allow(uint256,uint256,uint256,uint32)

the first 2 (aaaaaad1 and 00000000) sound really constructed and I do not think there is real world impact with these. And even 6e can IMHO be ignored. Wonder where your observation of :

> high number of collisions in the space

is coming from.

**Comment by prusnak** [comment]
> is coming from.

I looked at `00000000` and expected the rest is the same :-)

**Comment by prusnak** [comment]
Do you think it's possible to create something like a curated list where all collisions are resolved? That means no signature collisions (i.e. no entries for 00000000, 0000006e, etc.) and also where you pick the better value from `transfer(address _to,uint256 _value);transfer(address to,uint val)`? Or this is just too much effort?

**Comment by ligi** [comment]
hehe - no - don't judge a list by the first entry ;-)

I think 00000000 is constructed - some context: 
https://github.com/pipermerriam/ethereum-function-signature-registry/issues/30#issuecomment-385015547
http://swende.se/blog/Blockhash-Refactor.html

**Comment by ligi** [comment]
I think currently it is possible to curate a list that resolves the collisions - but not sure how future compatible this is ..

**Comment by prusnak** [comment]
I will check how many collisions there will be left if I treat arguments `_to` and `to` like they are the same. Also maybe creating a small dictionary which says that `val` and `value` are the same could help.

Another question: is `uint` and `uint256` the same?

**Comment by ligi** [comment]
> Another question: is uint and uint256 the same?

yes - uint *must* be resolved to uint256 for the function selector

> uint, int: synonyms for uint256, int256 respectively. For computing the function selector, uint256 and int256 have to be used.

https://solidity.readthedocs.io/en/develop/abi-spec.html

PS: the collisions I showed above are without parameter names - with parameter names there are more - currently I am not sure about the parameter names - hope to be able to discuss this in about 2 weeks in person with @holiman
